wire-ios-data-model/Source/Model/Message/ZMAssetClientMessage+Ephemeral.swift

```diff
@@ -65,16 +65,18 @@ extension ZMAssetClientMessage {
     }
 
     @discardableResult @objc public override func startDestructionIfNeeded() -> Bool {
-
-        if imageMessageData != nil && !hasDownloadedFile {
-            return false
-        } else if fileMessageData != nil
+        let imageNotDownloaded = imageMessageData != nil && !hasDownloadedFile
+        let fileHasNoUploadState = fileMessageData != nil
             && underlyingMessage?.assetData?.hasUploaded == false
-            && underlyingMessage?.assetData?.hasNotUploaded == false {
+            && underlyingMessage?.assetData?.hasNotUploaded == false
+        let hasOtherUserSender = sender == nil || (sender?.isSelfUser == false && managedObjectContext?.zm_isUserInterfaceContext == true)
+        let destructionStartCondition = !(fileHasNoUploadState || (hasOtherUserSender && imageNotDownloaded))
+        if destructionStartCondition {
+            return super.startDestructionIfNeeded()
+        }
```

Comment:
This is quite tricky to read, perhaps we can break it up a bit more:

```swift
if selfIsSender && !isAssetUploaded {
  return false
}

if !selfIsSender && !isAssetDownloaded {
  return false
}

return super.startDestructionIfNeede()

```

Reply:
It's not the same logic. Its checking the other sender instead not the self, Self sender logic was broken. Also the file upload state is checked regardless of the sender.

Reply:
Sure, but could it still be rewritten in a clearer way? I'm still struggling to understand condition.